File: CVNMS/src/nms.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <opencv2/opencv.hpp>
using namespace std;
using namespace cv;

// iou 两个矩形的交叉面积所占比例
// 0 - 1; 0 无交叉, 1 完全重叠

double IOU(const Rect& r1, const Rect& r2)
{
    int x1 = std::max(r1.x, r2.x);
    int y1 = std::max(r1.y, r2.y);
    int x2 = std::min(r1.x+r1.width, r2.x+r2.width);
    int y2 = std::min(r1.y+r1.height, r2.y+r2.height);
    int w = std::max(0, (x2-x1+1));
    int h = std::max(0, (y2-y1+1));
    double inter = w * h;
    double o = inter / (r1.area() + r2.area() - inter);
    return (o >= 0) ? o : 0;
}

void nms(vector<Rect>& proposals, const double nms_threshold)
{
    vector<int> scores;
    // 传入scores ,这里为了检测算法，矩形的面积当做scores
    for(auto i : proposals) scores.push_back(i.area());
    //矩形索引 {0,1,2,3,4,...}
    vector<int> index;
    for(int i = 0; i < scores.size(); ++i){
        index.push_back(i);
    }
    //根据scores从大到小排序后的矩形索引
    //{12,4,5,9,...}
    sort(index.begin(), index.end(), [&](int a, int b){
        return scores[a] > scores[b];
    }); 
    //每一个矩形进行标记是否要被删除
    vector<bool> del(scores.size(), false);
    for(size_t i = 0; i < index.size(); i++){
        if( !del[index[i]]){
            //选择score最高的一个(一般不会被删除)
            for(size_t j = i+1; j < index.size(); j++){
                //从剩余（i+1)个矩形中，依次和选择的矩形计算交叉面积比;
                //如果交叉面积的比值小于等于 设置的阈值 则证明两个矩形的不像似可以保留
                //否则被删除,被删除的矩形的index值被标记为true
                if(IOU(proposals[index[i]], proposals[index[j]]) > nms_threshold){
                    del[index[j]] = true;
                }
            }
        }
    }
    //将非true的index的矩形为新的备选矩形
    vector<Rect> new_proposals;
    for(const auto i : index){
        if(!del[i]) new_proposals.push_back(proposals[i]);
    }
    proposals = new_proposals;
}

void test_nms(char *path)
{
    Mat img = imread(path);
    Mat img2 = img.clone();

    // prepare data
    vector<Rect> proposals;
    Point2i origin(20, 20);
    int w = 100;
    int h = 200;
    for (auto i : {0,1,2,3,4,5}) proposals.push_back(Rect(origin.x + 10*i, origin.y + 20*i, w - 5*i, h-7*i));
    origin.x = 200;
    origin.y = 400;
    w = 300;
    h = 120;
    for (auto i : {0,1,2,3,4,5}) proposals.push_back(Rect(origin.x + 10*i, origin.y + 20*i, w - 5*i, h-7*i));

    proposals.push_back(Rect(400, 200, 386, 239));
    proposals.push_back(Rect(200, 120, 186, 209));
    proposals.push_back(Rect(310, 376, 286, 139));

    for(auto rect : proposals) rectangle(img, rect, Scalar(0, 255, 255), 2);
    imshow("before", img);
    // nms
    nms(proposals, 0.1);
    for(auto rect : proposals) rectangle(img2, rect, Scalar(0, 255, 255), 2);
    imshow("after", img2);
    waitKey(0);
    destroyAllWindows();
}
int main(int argc ,char **argv)
{
    test_nms(argv[1]);
    return 0;
}

File: CVNMS/src/nms.py
# nms python
import cv2 as cv
import numpy as np
import sys
def py_nms(dets, thresh):
    """Pure Python NMS baseline."""
    #x1、y1、x2、y2、以及score赋值
    x1 = dets[:, 0]
    y1 = dets[:, 1]
    x2 = dets[:, 2]
    y2 = dets[:, 3]
    scores = dets[:, 4]

    #每一个候选框的面积
    areas = (x2 - x1 + 1) * (y2 - y1 + 1)
    #order是按照score降序排序的
    order = scores.argsort()[::-1]
    keep = []
    while order.size > 0:
        i = order[0]
        keep.append(i)
        #计算当前概率最大矩形框与其他矩形框的相交框的坐标
        xx1 = np.maximum(x1[i], x1[order[1:]])
        yy1 = np.maximum(y1[i], y1[order[1:]])
        xx2 = np.minimum(x2[i], x2[order[1:]])
        yy2 = np.minimum(y2[i], y2[order[1:]])

        #计算相交框的面积
        w = np.maximum(0.0, xx2 - xx1 + 1)
        h = np.maximum(0.0, yy2 - yy1 + 1)
        inter = w * h
        #计算重叠度IOU：重叠面积/（面积1+面积2-重叠面积）
        ovr = inter / (areas[i] + areas[order[1:]] - inter)

        #找到重叠度不高于阈值的矩形框索引
        inds = np.where(ovr <= thresh)[0]
        #将order序列更新，由于前面得到的矩形框索引要比矩形框在原order序列中的索引小1，所以要把这个1加回来
        order = order[inds + 1]
    return keep

def genreateproposal(rect,proposal,n):
    for i in range(1,n+1):
        x1 = rect[0]+i*10
        y1 = rect[1]+i*10
        x2 = rect[2]+i*5
        y2 = rect[3]+i*5
        score = (x2 - x1 + 1) * (y2 - y1 + 1)
        proposal.append([x1,y1,x2,y2,score])
    return proposal
def prepareproposal(path):
    img =  cv.imread(path)
    imgmultiproposals = img.copy()
    proposals = []
    init_react1 = [10,10,200,300]
    # 4 个
    proposals = genreateproposal(init_react1,proposals,4)

    init_react2 = [100,120,300,500]

    proposals = genreateproposal(init_react2,proposals,5)

    proposals = np.array(proposals)

    for rec in proposals:
        cv.rectangle(imgmultiproposals,(rec[0],rec[1]), (rec[2],rec[3]),(0,0,255)) 
    cv.imshow("before",imgmultiproposals)

    keepindex = py_nms(proposals,0.1)
    imgnms = img.copy()
    for i in keepindex:
        rec = proposals[i]
        cv.rectangle(imgnms,(rec[0],rec[1]), (rec[2],rec[3]),(0,0,255)) 
    cv.imshow("after",imgnms)

    cv.waitKey(0)
    cv.destroyAllwindows()

if __name__ == "__main__":
    prepareproposal(sys.argv[1])